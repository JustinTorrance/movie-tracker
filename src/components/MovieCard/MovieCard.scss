.movie-card {
  border: 2px solid black;
  height: 400px;
  width: 250px;
  margin: 5px;
  transition: 0.3s;
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
}

img {
  width: 50px;
  height: 100px;
}

