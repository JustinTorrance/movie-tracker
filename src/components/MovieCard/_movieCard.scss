.movie-card {
  border: 2px solid black;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 400px;
  width: 250px;
  margin: 5px;
  transition: 0.3s;
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  position: relative;
  overflow: hidden;

  .dark-background {
    background-color: rgb(40, 41, 53);
    height: 400px;
    width: 250px;
    opacity: .03;
    transition: all .3s ease;
    position: absolute;
  }

  .movie-title-container {
    color: white;
    transform: translateY(-400px);
    transition: all .3s ease;
    margin: 8px 10px;
  }

  .overview {
    color: white;
    position: relative;
    transform: translateY(400px);
    transition: all .3s ease;
    margin: 10px;
    max-height: 200px;
    overflow: scroll;
  }

  &:hover .dark-background {
    cursor: pointer;
    opacity: .90;
  }

  &:hover .overview {
    transform: translateY(0px);
  }

  &:hover .movie-title-container {
    transform: translateY(0px);
  }
}

img {
  width: 50px;
  height: 100px;
}

.fa-star {
  font-size: 1.9em;
  padding: 8px;
  transition: all .3s ease;
  top: 0;
  right: 0;
  position: absolute;
  z-index: 2;

  &:hover {
    cursor: pointer;
    color: gold;
  }
}

.favorite {
  color: gold;
}

.movie-title {
  font-family: $regular-font;
  font-weight: 500;
  margin: 0px;
  margin-bottom: 10px;
  max-width: 85%;
  font-size: 2em;
}

.release-date-label,
.rating-label {
  display: inline-block;
  font-weight: 300;
  font-style: italic;  
  margin: 5px auto;
}

.movie-release-date, 
.movie-rating {
  display: inline-block;
  margin: 0px;
  margin-left: 5px;
  font-size: 1.2em;
  font-family: $title-font;
}
