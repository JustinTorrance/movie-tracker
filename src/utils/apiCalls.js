export const fetchMovies = async (url) => {
  try {
    const response = await fetch(url)
    const data = await response.json()
    return data 
  } catch(error) {
    throw new Error()
  }
}