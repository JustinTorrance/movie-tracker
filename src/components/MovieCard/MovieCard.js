import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import * as API from '../../utils/apiCalls'
import { connect } from 'react-redux'

export class MovieCard extends Component {
  constructor(){
    super()
    this.state = {
      favorite: false
    }
  }

  async componentDidMount() {
    const currentFavorites = await API.getFavorites(this.props.user_id)
    const found = currentFavorites.data.find((faveMovie) => {
      return faveMovie.movie_id === this.props.movie.id 
    })

    if (found) {
      this.setState({
        favorite: true
      })
    } else {
      this.setState({
        favorite: false
      })
    } 
  }

  toggleFavorite = () => {
    const { user_id, movie } = this.props
    if (this.state.favorite) {
      API.deleteFavorite(user_id, movie.id) 
    } else {
      API.addFavorite(movie.id, user_id, movie) 
    }
    this.setState({favorite: !this.state.favorite})
  }

  render() {
    const { movie } = this.props
    return (
    <article className='movie-card' style={{ backgroundImage: `url(${movie.poster_path})` }}>
      <i onClick={this.toggleFavorite} className={`fas fa-star ${this.state.favorite && 'favorite'}`}></i>    
      <div className='dark-background'></div>
      <div className='movie-title-container'>
        <h2 className='movie-title'>{movie.title}</h2>
        <p className='release-date-label'>Released:</p>
        <p className='movie-release-date'>{movie.release_date}</p>
        <p className='rating-label'>Avg. Rating:</p>
        <p className='movie-rating'>{movie.vote_average}</p>
      </div>
      <p className='overview'>{movie.overview}</p>
    </article>
    )
  }
} 

export const mapStateToProps = (state) => ({
  user_id: state.user.id
})

MovieCard.propTypes = {
  movie: PropTypes.object.isRequired,
  user_id: PropTypes.number.isRequired
}

export default connect(mapStateToProps)(MovieCard)
