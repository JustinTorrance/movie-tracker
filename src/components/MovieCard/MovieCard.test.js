import React from 'react'
import { shallow } from 'enzyme'
import MovieCard from './MovieCard'
import { mapStateToProps } from './MovieCard'
import { deleteFavorite, addFavorite, getFavorites } from '../../utils/apiCalls'
import { isRegExp } from 'util';

jest.mock('../../utils/apiCalls')

describe('MovieCard', () => {
  let mockMovie;
  let wrapper;

  beforeAll(() => {
     mockMovie = {
      id: 1,
      title: 'Titanic',
      year: 1999,
      rating: 9,
      posterPic: 'https://image.tmdb.org/t/p/w500///',
      backdropPic: 'https://image.tmdb.org/t/p/w500///',
      overview: 'great movie',
      genres: 'drama',
      runtime: 50
    }

    wrapper = shallow(
      <MovieCard movie={mockMovie} />
    );

  })

  describe('MovieCard Component', () => {
    it('should match the snapshot with all data passed in correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should call getFavorites on componentDidMount', () => {
      const mockUser_Id = 7
      getFavorites.mockImplementation(() => mockUser_Id)
  
      const wrapper = shallow(<MovieCard movie={mockMovie} />, { disableLifecycleMethods: true })
          wrapper.instance().componentDidMount()
          expect(getFavorites).toHaveBeenCalled()
  
    })

    it('should call toggleFavorite when deleting a favorite', () => {
      deleteFavoriteMock = jest.fn()
      const spy = spyOn(wrapper.instance(), 'toggleFavorite')
      wrapper.instance().forceUpdate()
      const mockEvent = deleteFavoriteMock 
      wrapper.find().simulate('change', mockEvent)
      expect(spy).toHaveBeenCalled()
    })

    it('should call toggleFavorite when adding a favorite', () => {


    })

    it('updates sate when toggleFavorite is called', () => {
      const wrapper = shallow(<MovieCard movie={mockMovie} toggleFavorite={toggleFavorite} />)
      const mockEvent = { favorite: true }
      wrapper.instance().toggleFavorite(mockEvent)
      expect(wrapper.state('favorite')).toBe(true)
    })
 


  describe('mapStateToProps', () => {
    it('should return with a user id', () => {

      const mockObject = {
        movie: {},
        user_id: 7
      }
      
      const mockUser_Id =  mockObject.user_id

      const expected = {
        user_id: mockUser_Id
      }

      const mockState = {
        user_id: mockUser_Id
      }

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })
  
  })
})