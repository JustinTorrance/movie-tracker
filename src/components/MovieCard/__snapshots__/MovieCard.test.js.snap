// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard MovieCard Component should match the snapshot with all data passed in correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieCard
    movie={
      Object {
        "backdropPic": "https://image.tmdb.org/t/p/w500///",
        "genres": "drama",
        "id": 1,
        "overview": "great movie",
        "posterPic": "https://image.tmdb.org/t/p/w500///",
        "rating": 9,
        "runtime": 50,
        "title": "Titanic",
        "year": 1999,
      }
    }
    user_id={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <i
          className="fas fa-star false"
          onClick={[Function]}
        />,
        <div
          className="dark-background"
        />,
        <div
          className="movie-title-container"
        >
          <h2
            className="movie-title"
          >
            Titanic
          </h2>
          <p
            className="release-date-label"
          >
            Released:
          </p>
          <p
            className="movie-release-date"
          />
          <p
            className="rating-label"
          >
            Avg. Rating:
          </p>
          <p
            className="movie-rating"
          />
        </div>,
        <p
          className="overview"
        >
          great movie
        </p>,
      ],
      "className": "movie-card",
      "style": Object {
        "backgroundImage": "url(undefined)",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "fas fa-star false",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "dark-background",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2
              className="movie-title"
            >
              Titanic
            </h2>,
            <p
              className="release-date-label"
            >
              Released:
            </p>,
            <p
              className="movie-release-date"
            />,
            <p
              className="rating-label"
            >
              Avg. Rating:
            </p>,
            <p
              className="movie-rating"
            />,
          ],
          "className": "movie-title-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Titanic",
              "className": "movie-title",
            },
            "ref": null,
            "rendered": "Titanic",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Released:",
              "className": "release-date-label",
            },
            "ref": null,
            "rendered": "Released:",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "movie-release-date",
            },
            "ref": null,
            "rendered": null,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Avg. Rating:",
              "className": "rating-label",
            },
            "ref": null,
            "rendered": "Avg. Rating:",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "movie-rating",
            },
            "ref": null,
            "rendered": null,
            "type": "p",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "great movie",
          "className": "overview",
        },
        "ref": null,
        "rendered": "great movie",
        "type": "p",
      },
    ],
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <i
            className="fas fa-star false"
            onClick={[Function]}
          />,
          <div
            className="dark-background"
          />,
          <div
            className="movie-title-container"
          >
            <h2
              className="movie-title"
            >
              Titanic
            </h2>
            <p
              className="release-date-label"
            >
              Released:
            </p>
            <p
              className="movie-release-date"
            />
            <p
              className="rating-label"
            >
              Avg. Rating:
            </p>
            <p
              className="movie-rating"
            />
          </div>,
          <p
            className="overview"
          >
            great movie
          </p>,
        ],
        "className": "movie-card",
        "style": Object {
          "backgroundImage": "url(undefined)",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fas fa-star false",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "dark-background",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2
                className="movie-title"
              >
                Titanic
              </h2>,
              <p
                className="release-date-label"
              >
                Released:
              </p>,
              <p
                className="movie-release-date"
              />,
              <p
                className="rating-label"
              >
                Avg. Rating:
              </p>,
              <p
                className="movie-rating"
              />,
            ],
            "className": "movie-title-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Titanic",
                "className": "movie-title",
              },
              "ref": null,
              "rendered": "Titanic",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Released:",
                "className": "release-date-label",
              },
              "ref": null,
              "rendered": "Released:",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "movie-release-date",
              },
              "ref": null,
              "rendered": null,
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Avg. Rating:",
                "className": "rating-label",
              },
              "ref": null,
              "rendered": "Avg. Rating:",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "movie-rating",
              },
              "ref": null,
              "rendered": null,
              "type": "p",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "great movie",
            "className": "overview",
          },
          "ref": null,
          "rendered": "great movie",
          "type": "p",
        },
      ],
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
